Kernel/Net/IPv4Socket.cpp

```diff
@@ -130,6 +130,16 @@ bool IPv4Socket::can_read(FileDescription& description) const
         return can_accept();
     if (protocol_is_disconnected())
         return true;
+
+    // This is sort of a hack. Normally, it's the socket-related
```

Comment:
This is so ugly.. we should think of a better way to do this. :/
It feels like it should be up to the Layer 4 protocol (TCP or UDP) to trigger moving the socket to `SocketRole::Connected` state.
What do you think? cc @deoxxa

Reply:
What I don't actually understand is why the `SocketRole` is a property of the fd (I mean `FileDescription`) and not the socket itself. If it was a property of sockets, this wouldn't be a problem. And now that I think of it this way, changing the role from `Process` looks like a hack — the socket should track and change its own role, shouldn't it?

Reply:
It's because for local sockets, you have multiple `FileDescription` objects referring to the same `LocalSocket`.

Note that this code was originally written for `LocalSocket` only, before there was any IPv4 code. So maybe this design doesn't make total sense now :)

Reply:
A solution would be to move this to being specific to local sockets only. They are the only ones who should care about the FD's socket role I think.

Reply:
> It's because for local sockets, you have multiple `FileDescription` objects referring to the same `LocalSocket`.

Do you though? Maybe I've missed something, but AFAICS:
* each time `socket()` is called, a new `LocalSocket` is created
* `bind()` creates a dummy inode on the VFS and sets itself as that inode's socket:
https://github.com/SerenityOS/serenity/blob/bd6d2c08199c992a9ac1f7c7752b3d7b1bf67bb9/Kernel/Net/LocalSocket.cpp#L66-L67
* `connect()` opens this dummy inode, find the peer — the bound listening socket — and grabs a reference to it:
https://github.com/SerenityOS/serenity/blob/bd6d2c08199c992a9ac1f7c7752b3d7b1bf67bb9/Kernel/Net/LocalSocket.cpp#L102-L103

Reply:
Yes, then `accept()` grabs that queued socket from `Socket::m_pending` and creates a second `FileDescription` for it :)

Reply:
Ah, that makes sense — so a `LocalSocket` is shared between the "accepted" and "connected" `FileDescription`s.

> A solution would be to move this to being specific to local sockets only. They are the only ones who should care about the FD's socket role I think.

That sounds good. So I'll move the role to the `Socket`, make a getter accept an fd, and then make `LocalSocket` differentiate between two fds/roles. Right?

Reply:
I'm not sure what you mean. How do you know which `FileDescription` is which if the role is in `Socket`?

Reply:
I was thinking `LocalSocket` would keep a weak reference to (one of) the descriptions and compare them to what gets passed as an argument.

Reply:
Hmm, I guess that could work. So the `LocalSocket` would point to one of the two FD's, and it does a pointer comparison to see if it's the accepting or connecting FD calling.

Reply:
It turns out that due to FDs attaching/detaching themselves we don't even need an actual `WeakPtr`, we can just store/clear a raw pointer. I find explicit `WeakPtr`s to be more clear, but an overhead of an additional cell per FD would be unfortunate.